src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,395 @@
+  aux_symbol.is_static_lifetime=static_lifetime;
+
+  exprt object=aux_symbol.symbol_expr();
+  exprt aoe=address_of_exprt(object);
+  if(cast_needed)
+  {
+    aoe=typecast_exprt(aoe, target_expr.type());
+  }
+  code_assignt assign(target_expr, aoe);
+  assign.add_source_location()=loc;
+  init_code.copy_to_operands(assign);
+
+  input_recorder.record_input(aux_symbol.base_name, object);
+
+  return aoe;
+}
+
+/*******************************************************************\
+
+Function: symbol_factoryt::gen_nondet_init
+
+ Inputs:
+  expr -
+  loc -
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void symbol_factoryt::gen_nondet_init(
+  const exprt &expr,
+  const source_locationt &loc)
+{
+  const typet &type=ns.follow(expr.type());
+
+  if(type.id()==ID_pointer)
+  {
+    // dereferenced type
+    const pointer_typet &pointer_type=to_pointer_type(type);
+    const typet &subtype=ns.follow(pointer_type.subtype());
+
+    code_labelt set_null_label;
+    code_labelt init_done_label;
+
+    static size_t synthetic_constructor_count=0;
+
+    if(!assume_non_null)
+    {
+      auto returns_null_sym=c_declare_new_tmp_symbol(
+        init_code,
+        symbol_table,
+        loc,
+        c_bool_typet(1),
+        "pointer_is_null");
+
+      auto returns_null=returns_null_sym.symbol_expr();
+      auto assign_returns_null=
+          code_assignt(returns_null, c_get_nondet_bool(returns_null_sym.type));
+      assign_returns_null.add_source_location()=loc;
+      init_code.move_to_operands(assign_returns_null);
+
+      auto set_null_inst=
+        code_assignt(expr, null_pointer_exprt(pointer_type));
+      set_null_inst.add_source_location()=loc;
+
+      std::string fresh_label=
+        "post_synthetic_malloc_"+std::to_string(++synthetic_constructor_count);
+      set_null_label=code_labelt(fresh_label, set_null_inst);
+
+      input_recorder.record_input(returns_null_sym.base_name, returns_null);
+
+      init_done_label=code_labelt(fresh_label+"_init_done", code_skipt());
+
+      code_ifthenelset null_check;
+      exprt null_return=
+        zero_initializer(returns_null_sym.type, loc, ns, message_handler);
+      null_check.cond()=
+        notequal_exprt(returns_null, null_return);
+      null_check.then_case()=code_gotot(fresh_label);
+      init_code.move_to_operands(null_check);
+    }
+
+    exprt allocated=allocate_object(expr, subtype, loc, false);
+    exprt init_expr;
+    if(allocated.id()==ID_address_of)
+    {
+      init_expr=allocated.op0();
+    }
+    else
+    {
+      init_expr=dereference_exprt(allocated, allocated.type().subtype());
+    }
+    gen_nondet_init(init_expr, loc);
+
+    if(!assume_non_null)
+    {
+      init_code.copy_to_operands(code_gotot(init_done_label.get_label()));
+      init_code.move_to_operands(set_null_label);
+      init_code.move_to_operands(init_done_label);
+    }
+  }
+  else
```

Comment:
This patch is only intended to make pointers work. Structs and arrays need to be done, but I think it is best left to another patch.